building-with-ocean/compute-to-data/user-defined-parameters.md

```diff
@@ -0,0 +1,183 @@
+
+There 2 are types of parameters that asset publishers can support:
+
+* User defined parameters
+* Algorithm custom parameters
+
+### Publish a dataset that uses custom parameters
+
+The dataset publisher can support these parameters to allow filtering, querying on the published dataset. The additional parameters that facilitate this are called `User defined parameters`. The Provider combines the original asset URL and the entered parameter values into a new URL and then streams the response from the modified URL back to the buyer.
+
+#### Use case for user defined parameters
+
+For example, if the publisher has published an URL `https://example.com` which serves large size historical weather data from all over the world. If the publisher wants to allow buyers to filter the data based on location, type of data, etc., it is possible to do so using user defined parameters.
+
+Suppose the publisher defines the following 2 parameters:
+
+* `location`: A string indicating region code
+* `type`: A string indicating the type of weather data. It can be temperature/humidity/pressure.
+
+Suppose the buyer wants to download the temperature data in the region code `XYZ`. While downloading the data buyer enters the desired parameter values using ocean.py or ocean.js.
+
+The provider will decrypt the URL from the DDO published on-chain, construct a new URL with the additional parameters and finally stream data to the buyer.
+
+Internally, the new URL will be of the format `https://example.com/?location=XYZ&type=temperature`. The server hosting the data has to read these parameters and serve the appropriate data.
+
+The following steps will specify how the publisher can support additional parameters.
+
+#### Step 1: Create a service
+
+The below python script exposes a REST endpoint that takes two parameters, namely: `location` and `type`. Let's assume that the dataset publisher hosts the service at domain `example.com` along with HTTPS support. The publisher must ensure that the URL is accessible to Provider.
+
+The code snippet is only for demo purposes and not for production use.
+
+```python
+from flask import Flask, request
+def get_data(data_type: str, location: str):
+    '''
+        Add some business logic here to get
+        the required data with given parameters
+    '''
+    return {}
+@app.route('/', methods=['GET'])
+def serve_content():
+    args = request.args
+    data_type = args.get('type')
+    location = args.get('location')
+    result = get_data(data_type, location)
+    return result
+```
+
+#### Step 2: Publish dataset asset with compute service
+
+The publisher now must provide the file URL as `https://example.com` while publishing the asset, as shown in the below image.
+
+![](../../.gitbook/assets/compute-to-data-parameters-publish-dataset.png)
+
+To view a complete tutorial on publishing asset using Ocean Marketplace click [here](../../using-ocean-market/marketplace-publish-data-asset.md).
```

Comment:
```suggestion
For a complete tutorial on publishing asset using Ocean Marketplace read [our guide on publishing with Ocean Market](../../using-ocean-market/marketplace-publish-data-asset.md).
```

Reply:
In general, let's avoid saying "click here". It's nicer if the link actually describes the content.